servers/shareable-lists-api/src/database/mutations/ShareableList.ts

```diff
@@ -101,6 +102,100 @@ export async function createShareableList(
   return list;
 }
 
+/**
+ * Create a new List and add one or more items to it at the same time.
+ * @param context
+ * @param data
+ * @param userId
+ * @returns
+ */
+export async function createAndAddToShareableList(
+  context: BaseContext,
+  data: CreateAndAddToShareableListInput,
+  userId: number | bigint,
+): Promise<ShareableList> {
+  const { db, conn } = context;
+  const { itemData, ...listData } = data;
+  const listInput = {
+    ...listData,
+    // again, prisma's bigint making type safety hard...
+    userId: userId as any,
+    externalId: uuid(),
+  };
+  // check if the title already exists for this user
+  // TODO (@kschelonka): add this as a constraint using generated hash column
+  // For future modeling, if we really want to be parsimonious we can use that
+  // hash as the "externalId"
+  const titleExists = await db.list.count({
+    where: { title: listData.title, userId: userId },
+  });
+
+  if (titleExists) {
+    throw new UserInputError(
+      `A list with the title "${listData.title}" already exists`,
+    );
+  }
+  // The relational link depends on an auto-generated listId, so that will be
+  // pouplated inside the transaction
+  const itemInput: Array<Omit<CreateListItemDb, 'listId'>> = itemData.map(
+    (item, index) => {
+      validateItemId(item.itemId);
+      return {
+        ...item,
+        itemId: parseInt(item.itemId),
+        sortOrder: index,
+        externalId: uuid(),
+      };
+    },
+  );
+  // Create a list and populate it with items in one transaction
+  // If the item insertion fails, then the list will not be created.
+  const { items, list } = await conn.transaction().execute(async (trx) => {
+    await trx.insertInto('List').values(listInput).execute();
+    const list = await trx
+      .selectFrom('List')
+      .where('externalId', '=', listInput.externalId)
+      .select(ShareableListSelect)
+      .executeTakeFirstOrThrow();
+    const itemInserts: CreateListItemDb[] = itemInput.map((item) => ({
+      ...item,
+      listId: list.id,
+    }));
+    await trx.insertInto('ListItem').values(itemInserts).execute();
+    const items = await trx
+      .selectFrom('ListItem')
+      .where('listId', '=', list.id)
+      .select(ListItemSelect)
+      .orderBy('sortOrder', 'asc')
+      .execute();
+    return { items, list };
+  });
+  const completeList: ShareableList = {
+    ...list,
+    // bigint causes issues again...
+    userId: list.userId as unknown as bigint,
+    listItems: items as any, // nested bigints
+  };
+  //send event bridge event for shareable-list-created event type
+  sendEventHelper(EventBridgeEventType.SHAREABLE_LIST_CREATED, {
```

Comment:
_in the future_ we will prob want or need a bulk add event.

Reply:
Yes